File: Residential_Controller/python/includes/column.py
from .elevator import Elevator

class Column: 
    def __init__ (self, elevatorNumbers, numberFloors):
        self.elevatorList = []
        self.floorsList = []

#           ***** CREATING FLOORLIST *****
        for i in range(numberFloors):
            self.floorsList.append(i+1)

#           ***** CREATING ELEVATORLIST WITH OBJECT ELEVATOR *****
        for i in range(elevatorNumbers):
            elevator = Elevator(1, "", "elevator " + str(i+1))
            self.elevatorList.append(elevator)

    
    def requestElevator(self, requestedFloor, requestDirection):
        self.requestedFloor = requestedFloor
        self.requestDirection = requestDirection
        self.findElevator(requestedFloor, requestDirection)

    def requestFloor(self, elevator, destination):
        self.destination = destination
        self.elevator = self.elevatorList[0]
        self.elevatorList[0].moveD(destination)

#              ***** FIND BEST ELEVATOR FOR SCENARIO DEPENDING ON VARIABLES *****
    def findElevator(self, requestedFloor, requestDirection):
        for elevator in self.elevatorList:
            floorsDiff = abs(elevator.currentFloor - requestedFloor)
            if(elevator.currentFloor == requestedFloor and elevator.direction =="idle"):
                elevator.elevatorScore = 0 + floorsDiff
            elif(elevator.currentFloor < requestedFloor and requestDirection == elevator.direction and elevator.direction == "up"):
                elevator.elevatorScore = 100 + floorsDiff
            elif(elevator.currentFloor > requestedFloor and requestDirection == elevator.direction and elevator.direction == "down"):
                elevator.elevatorScore = 200 + floorsDiff
            elif(elevator.currentFloor == requestedFloor and elevator.direction == "idle"):
                elevator.elevatorScore = 300 + floorsDiff
            elif(elevator.currentFloor != requestedFloor and elevator.direction != requestDirection and elevator.direction != "idle"):
                elevator.elevatorScore = 400 + floorsDiff
            elif(elevator.currentFloor == requestedFloor and elevator.direction != requestDirection):
                elevator.elevatorScore = 500 + floorsDiff
            elif(elevator.currentFloor != requestedFloor and elevator.direction == requestDirection):
                elevator.elevatorScore = 600 + floorsDiff
            elif(elevator.direction == "idle"):
                elevator.elevatorScore = 700 + floorsDiff
        self.elevatorList.sort(key=lambda x: x.elevatorScore)
        bestElevator = self.elevatorList[0]
        bestElevator.move(requestedFloor)

File: Commercial_Controller/Java/commercial_controller/src/app/Column.java
package app;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Collections;

public class Column{
    public String name;
    public ArrayList<Integer> floorList;
    public ArrayList<Elevator> elevatorList;
  
    public Column(String name, int numberOfElevators,int numberOfFloors){
    this.name = name;
    this.elevatorList = new ArrayList<Elevator>();
    this.floorList = new ArrayList<Integer>();
      }
      public Elevator findElevator(int floorNumber, int requestedFloor, Column bestCol){
          int size = bestCol.elevatorList.size();
        for (var i = 0; i < size; i++){
            int floorsDiff = Math.abs(bestCol.elevatorList.get(i).currentFloor - floorNumber);
    
            if(requestedFloor >= 1 && floorNumber >= 1 ){
                if (bestCol.elevatorList.get(i).currentFloor == floorNumber) {
                    bestCol.elevatorList.get(i).score = floorsDiff;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "down"){
                    bestCol.elevatorList.get(i).score = floorsDiff + 100;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "idle" ){
                    bestCol.elevatorList.get(i).score = floorsDiff + 200;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "up") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 300;
                } else if (floorNumber > 1 && bestCol.elevatorList.get(i).currentFloor > floorNumber && bestCol.elevatorList.get(i).direction == "down") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 400;
                } else if (floorNumber > 1 && bestCol.elevatorList.get(i).currentFloor > floorNumber && bestCol.elevatorList.get(i).direction == "idle" ){
                    bestCol.elevatorList.get(i).score = floorsDiff + 500;
                } else if (floorNumber > 1 && bestCol.elevatorList.get(i).currentFloor > floorNumber && bestCol.elevatorList.get(i).direction == "up") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 600;
                } else if (floorNumber > 1 && bestCol.elevatorList.get(i).currentFloor < floorNumber){
                    bestCol.elevatorList.get(i).score = floorsDiff + 700;
                }
    
            } else {
                if( bestCol.elevatorList.get(i).direction == "idle" && bestCol.elevatorList.get(i).currentFloor == floorNumber ){
                    bestCol.elevatorList.get(i).score = floorsDiff;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "up" ){
                    bestCol.elevatorList.get(i).score = floorsDiff + 100;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "idle" ){
                    bestCol.elevatorList.get(i).score = floorsDiff + 200;
                } else if (floorNumber == 1 && bestCol.elevatorList.get(i).direction == "down") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 300;
                } else if (floorNumber < 1 && bestCol.elevatorList.get(i).currentFloor < floorNumber && bestCol.elevatorList.get(i).direction == "up") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 400;
                } else if (floorNumber < 1 && bestCol.elevatorList.get(i).currentFloor < floorNumber && bestCol.elevatorList.get(i).direction == "idle") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 500;
                } else if (floorNumber < 1 && bestCol.elevatorList.get(i).currentFloor < floorNumber && bestCol.elevatorList.get(i).direction == "down") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 600;
                } else if (floorNumber < 1 && bestCol.elevatorList.get(i).currentFloor > floorNumber) {
                    bestCol.elevatorList.get(i).score = floorsDiff + 700;
                } else if (floorNumber < 1 && bestCol.elevatorList.get(i).direction == "down") {
                    bestCol.elevatorList.get(i).score = floorsDiff + 800;
                }
            } 
        }
        
        //										*****SORT ELEVATOR LIST ARRAY FROM LOWEST SCORE TO HIGHEST*****
            

            Collections.sort(bestCol.elevatorList, new Comparator<Elevator>() {
                @Override
                public int compare(Elevator o1, Elevator o2) {
                    return o1.score.compareTo(o2.score);
                }
            });
            
            var bestElev = bestCol.elevatorList.get(0);
            bestElev.move(floorNumber, bestElev);
            bestElev.moveD(requestedFloor, bestElev);

            
                return bestElev;
    
      }
    
    }   
 